Dynamic load libraries for external functions

artP2 opened the pull request.

Grillo-0 commented on src/extern_library_helper.h:
```diff
@@ -0,0 +1,27 @@
+#ifndef __EXTERN_LIBRARY_HELPER_H__
+#define __EXTERN_LIBRARY_HELPER_H__
+
+typedef char* (*FunctionPointer) (char*);
```
We're using snake case

Grillo-0 commented on src/extern_library_helper.h:
```diff
@@ -0,0 +1,27 @@
+
+typedef struct {
+  int counter;
+  void* handle;
```
Maybe a better name would be `dl_handle`

Grillo-0 commented on src/extern_library_helper.h:
```diff
@@ -0,0 +1,27 @@
+  char** names;
+} LibraryDefinition;
+
+LibraryDefinition expose_library();
```
Maybe those "user facing" definitions could be on a separated header file

artP2 replied:
Moved to lib/library_api.h

Grillo-0 commented on lib/test_lib.c:
```diff
@@ -0,0 +1,35 @@
+#include "../src/extern_library_helper.h"
+#include <stdio.h>
+
+#define NUM_FUNCTIONS 1
+LibraryDefinition expose_library() {
+    char* names[NUM_FUNCTIONS] = {"day_of_week"};
+    LibraryDefinition lib = {NUM_FUNCTIONS, names};
```
I don't like that the user needs to keep track of the number of functions it implemented.

Maybe, instead of a counter, we could have a blank last entry, like this.

```c
    LibraryDefinition lib = {
        "day_of_week",
        0,
    }
```

artP2 replied:
I created a macro in which the user only needs to provide the functions' names and pointers, like this:
```c
DEFINE_LIBRARY(
    FUNCTION("day_of_week", (function_pointer) day_of_week)
    )
```

Grillo-0 replied:
I think the `FUNCTION` macro is not needed, it just declares a struct

artP2 replied:
It's not necessary, but I think it makes it more legible. Here's an example of what it would look like without this macro:
```c
DEFINE_LIBRARY(
    (extern_function){"day_of_week", (function_pointer) day_of_week}
);
```
What do you think?

Grillo-0 replied:
The casting is necessary? Also, don't declare the structs with typedef.

artP2 replied:
Now the user no longer has to cast (the macro does that), I've removed the typedef from the structs and I've also simplified the code by removing the function counter.

Grillo-0 commented on src/extern_library_helper.c:
```diff
@@ -0,0 +1,39 @@
+  for (int i = 0; i < lib_def.counter ; i++) {
+    ExternFunction fr = {
+      lib_def.names[i],
+      dlsym(handle, lib_def.names[i])
```
This needs more attention: 

What if the user creates the external library in a language that does [name mangling](https://en.wikipedia.org/wiki/Name_mangling), like C++?

The user will not know the symbol name of the functions they want to provide.

artP2 replied:
Now we don't need the function name.

Grillo-0 commented on subprojects/stdlib/library_api.h:
```diff
@@ -0,0 +1,14 @@
+
+#define DEFINE_LIBRARY(...)                                              \
+    struct extern_function* expose_library() {                           \
+        static struct extern_function lib[] = {__VA_ARGS__, {0, 0}};     \
```
To keep a bit more future proof

```suggestion
        static struct extern_function lib[] = {__VA_ARGS__, { }};     \
```

Grillo-0 commented on subprojects/vinumc/extern_library_helper.h:
```diff
@@ -0,0 +1,19 @@
+typedef char* (*function_pointer) (char* argv);
+
+struct extern_function {
+  char* name;
```
```suggestion
  char *name;
```

Grillo-0 commented on subprojects/vinumc/extern_library_helper.h:
```diff
@@ -0,0 +1,19 @@
+
+struct loaded_lib {
+  void* dl_handle;
+  struct extern_function* functions;
```
```suggestion
  void *dl_handle;
  struct extern_function *functions;
```

Grillo-0 commented on subprojects/vinumc/extern_library_helper.h:
```diff
@@ -0,0 +1,19 @@
+  struct extern_function* functions;
+};
+
+struct loaded_lib load_lib(char* name);
```
```suggestion
struct loaded_lib load_lib(char *name);
```

Grillo-0 commented on subprojects/vinumc/extern_library_helper.h:
```diff
@@ -0,0 +1,19 @@
+#ifndef __EXTERN_LIBRARY_HELPER_H__
+#define __EXTERN_LIBRARY_HELPER_H__
+
+typedef char* (*function_pointer) (char* argv);
```
```suggestion
typedef char* (*function_pointer) (char *argv);
```

Grillo-0 commented on subprojects/vinumc/extern_library_helper.c:
```diff
@@ -0,0 +1,28 @@
+  // load library
+  void *handle = dlopen(name, RTLD_LAZY);
+  if (!handle) {
+    // error
```
This is a todo? I fit it is plead add a `TODO: ` flag and explaint a bit more

Grillo-0 commented on subprojects/vinumc/extern_library_helper.c:
```diff
@@ -0,0 +1,28 @@
+
+  struct extern_function* (*expose_library) () = dlsym(handle, "expose_library");
+
+  struct extern_function* functions = expose_library();
```
Always place the `*` next to the variable name, not the type

Grillo-0 commented on subprojects/stdlib/stdlib.c:
```diff
@@ -0,0 +1,32 @@
+}
+
+DEFINE_LIBRARY(
+    FUNCTION("day_of_week", day_of_week)
```
I still don't think this macro is necessary, it only declares a struct and does a casting.

However, the casting of `day_of_week` is needed here because the return type is different from `function pointer`. One returns `char *` and the other returns `const char*`.

If you fix this inconsistency, the casting is not necessary, and the removal of the macro makes sense.

Grillo-0 commented on subprojects/stdlib/library_api.h:
```diff
@@ -1,7 +1,7 @@
 #define __LIBRARY_API_H__
 
-#include "../src/extern_library_helper.h"
+#include <extern_library_helper.h>
```
This doesn't look great on the diff, on the previous commit you added this line.

I think those first two commits are not divided right. Also, the second commit message doesn't make much sense for me.

Move the stdlib part from the first commit to the second, and do the opposite for the modifications on the vinumc. So the first commit would create a base for the creation of  stdlib

Grillo-0 commented on subprojects/stdlib/stdlib.c:
```diff
@@ -0,0 +1,32 @@
+        return "[ERROR: Invalid date format]";
+    }
+
+    // Zeller
```
What is Zeller?

Grillo-0 commented on subprojects/vinumc/meson.build:
```diff
@@ -26,6 +26,7 @@ common_srcs = [
   'eval.c',
+  'extern_library_helper.c',
 ]
 
```
Also, add to an include dir `extern_library_helper.h`, like vunit does. To expose it to the user when it installs vinumc

artP2 replied:
I have no experience with Meson, the way I did it is working but I'm pretty sure it's not the most idiomatic way. I'll need some help.

Grillo-0 commented on subprojects/vinumc/dry_bison.y:
```diff
@@ -16,6 +16,7 @@ int yylex();
 %token ARG_REF_ALL_ARGS
 %token PROGRAM
 %token SYMBOL
+%token FUNCTION
```
I see why you needed to do this, I don't like the fact that we need to do this XD.

One more thing that makes me want to stop relying on the bison token types and create our own. For now, let's leave it at that, just add a TODO for future work.

Grillo-0 commented on subprojects/vinumc/eval.h:
```diff
@@ -4,13 +4,20 @@
 
 struct scope_childs_t VEC_DEF(size_t);
 
+enum EntryType {Symbol, Function};
```
```suggestion
enum entry_type {
	ENTRY_SYMBOL,
	ENTRY_FUNCTION,
};
```

Grillo-0 commented on subprojects/vinumc/eval.h:
```diff
@@ -31,7 +38,13 @@ struct eval_ctx {
+
+struct libraries {
+	int counter;
+	char** names;
```
the `*` thing

artP2 replied:
sorry, bad habit.

Grillo-0 commented on subprojects/vinumc/ast.h:
```diff
@@ -15,7 +16,10 @@ struct ast_node_childs_t VEC_DEF(ast_node_id_t);
 struct ast_node {
 	int type;
-	char *text;
+	union {
+		char *text;
+		function_pointer fp;
+	};
```
Do we really need to add the function pointers to the AST? Just adding them to the scopes as another entry doesn't solve the problem?

artP2 replied:
It would be necessary to pass the scope to do_calls(). I don't know if we want that.

Grillo-0 replied:
I don't see it as a problem, it would have the positive of encapsulating this new feature to just one data structure.

It would make the AST ignorant of function types, it would be just symbols.

Also, fewer places to change in the future.

artP2 replied:
Done.

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -244,15 +268,51 @@ DO_CALLS_FUNC_SIGNATURE(do_calls) {
 			.node = 0,
 	}));
 
+	struct loaded_lib* loaded_libs = load_libs(ctx, 0, libs);
 	resolve_symbols(ctx, ast, 0, 0);
 	resolve_calls(ctx, ast, 0, 0);
 	do_calls(ast, out, 0);
+	unload_libs(loaded_libs, libs.counter);
```
are you using the `loaded_libs`?

artP2 replied:
I use it to free the dl handles after use

Grillo-0 commented on subprojects/vinumc/vin2dot.c:
```diff
@@ -140,20 +140,23 @@ int main(int argc, char *argv[]) {
 			return EXIT_SUCCESS;
 		case CMD_AST_AND_SCOPES:
-			eval(&ctx.eval_ctx, &ctx.ast, stderr);
+			eval(&ctx.eval_ctx, &ctx.ast, stderr, libs);
```
All of this should be added at the same time you change the `eval` signature, every commit must be compilable.

Grillo-0 commented on subprojects/vinumc/tests/library_test.c:
```diff
@@ -0,0 +1,33 @@
+#include <vunit.h>
```
Add the tests in the same commit that you add the feature that you're testing

Grillo-0 reviewed (changes requested):
Some first review comments.

The commit structure and message needs some work, it would be great if you describe more of each commit with a body, some of the time a single line doesn't do the trick.

artP2 committed:
subprojects/vinumc: Add a helper to handle external libraries
Add vinumc/extern_library.c and vinumc/extern_library.h which provide
structures and functions for loading external libraries using dlfcn.

artP2 commented:
My commits were quite messy, now it should be much more concise. But I'm not happy with library_api.h in the last commit, any suggestions?

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -244,15 +268,51 @@ DO_CALLS_FUNC_SIGNATURE(do_calls) {
+}
+
+//TODO isso ta feio
+struct loaded_lib *load_libs (struct eval_ctx *ctx, size_t curr_scope_id, struct libraries libs){
```
```suggestion
struct loaded_lib *load_libs(struct eval_ctx *ctx, size_t curr_scope_id, struct libraries libs){
```

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -244,15 +268,51 @@ DO_CALLS_FUNC_SIGNATURE(do_calls) {
 }
 
-void eval(struct eval_ctx *ctx, struct ast *ast, FILE *out) {
+void resolve_extern_functions (struct eval_ctx *ctx, size_t curr_scope_id, struct loaded_lib lib){
```
```suggestion
void resolve_extern_functions(struct eval_ctx *ctx, size_t curr_scope_id, struct loaded_lib lib){
```

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -201,10 +210,25 @@ DO_CALLS_FUNC_SIGNATURE(do_calls_call) {
+	}else{
+		for (size_t i = 0; i < args_node->childs.len; i++) {
+			do_calls(ast, out, VEC_AT(&args_node->childs, i));
+		}
```
```suggestion
	if (symbol_node->type == FUNCTION){
		size_t buffer_size = 1024;
		char *buffer = (char*) calloc(1, buffer_size);
		FILE *tmp_out = fmemopen(buffer, buffer_size, "w+");
	}
	for (size_t i = 0; i < args_node->childs.len; i++) {
		do_calls(ast, tmp_out, VEC_AT(&args_node->childs, i));
	}
	if (symbol_node->type == FUNCTION){
		fflush(tmp_out);
		fprintf(out, "%s", symbol_node->fp(buffer));
		fclose(tmp_out);
		free(buffer);
	}
```

Grillo-0 commented:
> My commits were quite messy, now it should be much more concise. But I'm not happy with library_api.h in the last commit, any suggestions?

I don't see a problem with the `library_api.h` in the last commit, it is part of the stdlib and on this commit you are creating the stdlib.

Grillo-0 commented on subprojects/vinumc/meson.build:
```diff
@@ -32,9 +32,22 @@ vinumc_srcs = common_srcs + [
+  include_directories : include_directories('.'),
+  link_with : extern_library_helper_lib
+)
+
```
why you're separating this code into a shared library?

Grillo-0 commented on subprojects/vinumc/extern_library_helper.c:
```diff
@@ -0,0 +1,24 @@
+    //TODO: deal with the error of not being able to open the library
+  }
+
+  struct extern_function* (*expose_library) () = dlsym(handle, "expose_library");
```
I think this could be a function pointer type, and also be exposed to the user in sseparate header file, along with `extern_function`.

Grillo-0 commented on subprojects/vinumc/extern_library_helper.c:
```diff
@@ -0,0 +1,24 @@
+#include "extern_library_helper.h"
```
Maybe `extern_library` is a better name? The `helper` part dos not give that much more information

Grillo-0 commented on subprojects/vinumc/vinumc.c:
```diff
@@ -33,10 +33,19 @@ int main(int argc, char **argv) {
+				i--;
+			}
 		} else {
 			yyin = fopen(arg, "r");
```
Maybe give this its own commit, one commit create the infrastructure and the other exposes the created infrastructure to the user

Grillo-0 commented on subprojects/vinumc/eval.h:
```diff
@@ -4,13 +4,23 @@
 
+enum entry_type {
+	ENTRY_SYMBOL,
+	ENTRY_FUNCTION,
```
Maybe it would be better to change the name to  `ENTRY_INTERNAL` and `ENTRY_EXTERNAL`. A function is still a symbol, the new names be more meaningful.

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -201,10 +210,25 @@ DO_CALLS_FUNC_SIGNATURE(do_calls_call) {
-	for (size_t i = 0; i < args_node->childs.len; i++) {
-		do_calls(ast, out, VEC_AT(&args_node->childs, i));
+	if (symbol_node->type == FUNCTION){
+		size_t buffer_size = 1024;
```
Is better to use define for this

Grillo-0 commented on subprojects/vinumc/extern_library_helper.h:
```diff
@@ -0,0 +1,19 @@
+#ifndef __EXTERN_LIBRARY_HELPER_H__
+#define __EXTERN_LIBRARY_HELPER_H__
+
+typedef char* (*function_pointer) (char *argv);
```
`function_pointer` is too generic, maybe `ext_function_t` is a better one (or something better XD)?

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -201,10 +210,25 @@ DO_CALLS_FUNC_SIGNATURE(do_calls_call) {
+		fflush(tmp_out);
+		fprintf(out, "%s", symbol_node->fp(buffer));
+		fclose(tmp_out);
+		free(buffer);
```
I don't know exactly why this is being done. Why are you creating a buffer backed file and also using it as the args buffer to the external function?

artP2 replied:
Check if you understand with the comments I just made in the code, if not I should improve them.

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -244,15 +268,51 @@ DO_CALLS_FUNC_SIGNATURE(do_calls) {
+	}
+}
+
+//TODO isso ta feio
```
remove this XD, it's not that bad

Grillo-0 replied:
why do you think this is ugly?

Grillo-0 commented on subprojects/vinumc/vin2dot.c:
```diff
@@ -140,20 +140,23 @@ int main(int argc, char *argv[]) {
 
+	// TODO
+	struct libraries libs = {0, NULL};
+
```
`TODO` what?

Grillo-0 commented on subprojects/vinumc/vinumc.c:
```diff
@@ -33,10 +33,19 @@ int main(int argc, char **argv) {
 	setlocale(LC_ALL, "");
 
 	FILE *out = stdout;
+	struct libraries libs = {0, NULL};
```
I think its better to add this a new item on the vinum_ctx

artP2 replied:
Done, but why do we pass the individual entries to eval instead of the entire ctx struct?

Grillo-0 replied:
I think this way states what things inside the `ctx` are going to be read and modified. Currently, is all the struct fields, but in the future probably not. But meh, is not that critical.

Also, adding another argument is less intrusive than changing the whole arguments.

Grillo-0 commented on subprojects/vinumc/vinumc.c:
```diff
@@ -33,10 +33,19 @@ int main(int argc, char **argv) {
+			}
+			if (i != argc){
+				i--;
+			}
```
A bit sketchy to use the argv pointer to hold the library names.

It's better to copy the lib names to a new allocated array and make the `libs` object the **owner** ~:crab:~ of these strings.

Grillo-0 commented on subprojects/stdlib/meson.build:
```diff
@@ -0,0 +1,14 @@
+stdlib = shared_library(
+  'stdlib',
+  stdlib_srcs,
+  dependencies : extern_library_helper_dep,
```
Oh, now I see why you create the shared lib. You can create the dep object with only the header file, so you don't need to create a new shared lib.

```meson
my_dependency = declare_dependency(
  include_directories: include_directories('path/to/header')
)

```

artP2 replied:
Done.

Grillo-0 commented on subprojects/vinumc/tests/library_test.c:
```diff
@@ -0,0 +1,33 @@
+	vunit_run_vinumc_ok(ctx,
+		"[day_of_week 2024-09-26]\n",
+		&out,
+		"--with","subprojects/stdlib/libstdlib.so" , NULL
```
hum, maybe create an external lib just for testing inisde the vinumc project. If you want to test the stdlib behavior (which we probably should) create a test dir on the stdlib project.

artP2 replied:
that was the purpose of the stdlib of this pr

Grillo-0 replied:
I see, but I don't like that vinumc tests depend on another project.

Create another lib inside `test` for isolating the two projects. For now, it could be just a dumb function that returns some string, for the current tests it will suffice. 

The stdlib part of this PR will be the starting point.

Grillo-0 replied:
this is still an issue

Grillo-0 reviewed (changes requested):
About the commit messages, **much better**, just one nitpick, capitalize the subject lines, and also you forgot to capitalize the body of the first commit

Grillo-0 reviewed (changes requested):
Also, documentation stuff.
 
Add an entry to the list of subprojects on the `./README.md` and also create a separate `README.md` for stdlib.

Also, we need a way to document the API used for creating an external lib. Maybe create a file on the docs folder showing how to do it?

Grillo-0 reviewed (changes requested):
Also, let's start using the semantic versioning, bump vinumc version to 0.1.0.

artP2 committed:
docs/library_development.md: Add instructions on library development

artP2 committed:
subprojects/vinumc: Handle external libraries
Gives the compiler the ability to load and execute external functions.

Fixes: #50
Co-developed-by: Jean Patrick <jeanjpnm@usp.br>

artP2 committed:
subprojects/vinumc: Add "with" flag to load extern libraries
Add the flag "--with" allowing the user to specify a extern library.

artP2 committed:
vinumc: Bump version to 0.1.0

artP2 committed:
vinumc: Extend argument parsing to support multi-argument flags
This adds FLAG_MULTI_ARGUMENT to allow specifying arguments repeatedly.

Grillo-0 commented on subprojects/vinumc/include/extern_library.h:
```diff
@@ -0,0 +1,19 @@
+};
+
+struct loaded_lib load_lib(char *name);
+void unload_lib(struct loaded_lib lib);
```
use tabs instead of spaces, the tab width should be 8 spaces

Grillo-0 replied:
Library developers don't need to know about these functions, add them in a separate internal header (outside the `include` directory)

Grillo-0 commented on subprojects/vinumc/include/extern_library.h:
```diff
@@ -0,0 +1,19 @@
+#ifndef __EXTERN_LIBRARY_HELPER_H__
+#define __EXTERN_LIBRARY_HELPER_H__
```
the header guard tag doesn't match the file name

Grillo-0 commented on subprojects/stdlib/stdlib.c:
```diff
@@ -0,0 +1,32 @@
+}
+
+DEFINE_LIBRARY(
+    {"day_of_week", day_of_week}
```
I don't  think this macro is necessary, just create the `expose_library` directly

Grillo-0 commented on subprojects/vinumc/tests/meson.build:
```diff
@@ -4,6 +4,7 @@ tests = [
   'output_flag_test.c',
   'simple_test.c',
+  'library_test.c'
 ]
```
insert the new test in alphabetical order

Grillo-0 commented on subprojects/vinumc/meson.build:
```diff
@@ -32,9 +32,14 @@ vinumc_srcs = common_srcs + [
 vinumc = executable(
   'vinumc',
   vinumc_srcs,
+  dependencies : extern_library_dep,
```
the `executable`, has an `include_directories` argument itself

Grillo-0 commented on subprojects/vinumc/meson.build:
```diff
@@ -32,9 +32,14 @@ vinumc_srcs = common_srcs + [
   'vinumc.c',
 ]
 
+extern_library_dep = declare_dependency(
```
make the name more generic, the dependency is the whole include dir

Grillo-0 replied:
Also use `install_header` to install the header when installing the vinumc, install in the vinumc subdir

artP2 replied:
No clue on install_header for a declared dependency.

Grillo-0 commented on subprojects/vinumc/ast.h:
```diff
@@ -5,6 +5,7 @@
 #include <stdint.h>
 #include <stdio.h>
 
+#include <extern_library.h>
```
use the "" way of including this header, let's use te <> include way only for system libs

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -123,36 +127,42 @@ RESOLVE_FUNC_SIGNATURE(resolve_calls_call) {
+		} else if (symbol_info->type == ENTRY_EXTERNAL) {
+			struct ast_node *symbol_node = &VEC_AT(&ast->nodes, VEC_AT(&ast_node.childs, 0));
+			symbol_node->type = FUNCTION;
+    }
```
use tabs

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -231,28 +265,74 @@ DO_CALLS_FUNC_SIGNATURE(do_calls) {
+	// for (int i = 0; i < libraries->counter; i++){
+	// 	loaded_libs[i] = load_lib(libraries->names[i]);
+	// 	resolve_extern_functions(ctx, loaded_libs[i]);
+	// }
```
remove this

Grillo-0 commented on subprojects/vinumc/eval.h:
```diff
@@ -4,13 +4,23 @@
+	union {
+		int ast_node_id;
+		extern_function_pointer fp;
+	};
```
make the union named eith `as`, so we explicitly access a union

Grillo-0 commented on subprojects/vinumc/eval.h:
```diff
@@ -4,13 +4,23 @@
+	enum entry_type type;
+	union {
+		int ast_node_id;
+		extern_function_pointer fp;
```
`s/fp/func`

Grillo-0 commented on subprojects/vinumc/vinumc.c:
```diff
@@ -107,6 +111,19 @@ static void parse_cmdline(const int argc, char **argv, struct ctx *ctx, struct f
+					}
+					*f->ref_as.str_array = calloc(count + 1, sizeof(char *));
+					for (size_t i = 0; i < count; i++) {
+						(*f->ref_as.str_array)[i] = strdup(argv[optind - count + i]);
```
why the parenthesis around this line?

Grillo-0 commented on subprojects/vinumc/vinumc.c:
```diff
@@ -184,6 +184,13 @@ int main(int argc, char **argv) {
 		},
+		{
+			.name = "with", .short_name = 'w',
+			.help_desc = "Set a extern library to load",
```
Explain that multiple libraries can be added

Grillo-0 commented on subprojects/vinumc/vinumc.c:
```diff
@@ -107,6 +111,19 @@ static void parse_cmdline(const int argc, char **argv, struct ctx *ctx, struct f
 			case FLAG_BOOLEAN:
 				*f->ref_as.boolean = true;
 				break;
+			case FLAG_MULTI_ARGUMENTS:
+				{
+					size_t count = 0;
+					while (optind < argc && argv[optind][0] != '-') {
+						count++;
+						optind++;
+					}
+					*f->ref_as.str_array = calloc(count + 1, sizeof(char *));
+					for (size_t i = 0; i < count; i++) {
+						(*f->ref_as.str_array)[i] = strdup(argv[optind - count + i]);
+					}
```
Maybe this could be simplified by making the user specify each new arg using the flag again, like the gcc `-l` flag.

Grillo-0 replied:
This also would remove the problem of using this type of flag right before the file to  be compiled. Right now the file would be treated as an argument of the flag.

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -295,18 +296,21 @@ void resolve_extern_functions(struct eval_ctx *ctx, struct loaded_lib lib){
 }
 
-struct loaded_lib *load_libs(struct eval_ctx *ctx, char *libraries){
+struct loaded_lib *load_libs(struct eval_ctx *ctx, char **libraries){
```
squash this part to the commit that add this feature

Grillo-0 commented on subprojects/vinumc/vinumc.c:
```diff
@@ -184,6 +184,13 @@ int main(int argc, char **argv) {
+			.name = "with", .short_name = 'w',
+			.help_desc = "Set a extern library to load",
+			.placeholder_name = "libraries",
+			//TODO: load more than one
```
remove this todo, you're already supporting it

Grillo-0 commented on docs/library_development.md:
````diff
@@ -0,0 +1,19 @@
+      )
+  ```
+
+ 4. Compile it as a shared library and you're good to go.
````
show the command to use the shared library

Grillo-0 commented on docs/library_development.md:
````diff
@@ -0,0 +1,19 @@
+
+ 2. Define functions, they should accept a parameter of type char* and return a char*. 
+
+ 3. Register your functions using the macro `DEFINE_LIBRARY` this way:
+   ``` c
+  DEFINE_LIBRARY(
+      {"function_name", function_pointer},
+      {"function_name2", function_pointer2}
+      )
+  ```
````
change this part

Grillo-0 commented on docs/library_development.md:
```diff
@@ -0,0 +1,19 @@
+ 
+Follow these steps to create a Vinum library:
+
+ 1. Include the `library_api.h`.
```
I don't think the library developer should need to include a header from the stdlib, the developer should only need the header from vinumc

artP2 committed:
subprojects/vinumc: Add test library
This library is used for the unit tests.

Grillo-0 commented on subprojects/vinumc/extern_library.c:
```diff
@@ -0,0 +1,24 @@
+
+void unload_lib(struct loaded_lib lib) {
+  dlclose(lib.dl_handle);
+}
```
fix indentation, all tabs please

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -201,14 +213,38 @@ DO_CALLS_FUNC_SIGNATURE(do_calls_call) {
 		return;
 	}
 
+
```
Remove this new line

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -201,14 +213,38 @@ DO_CALLS_FUNC_SIGNATURE(do_calls_call) {
+		// writes the returns of the arguments calls to a temporary buffer,
+		// so any nested call will be resolved normally
+		char *buffer = (char*) calloc(1, EXTERN_CALL_BUFFER_SIZE);
+		FILE *tmp_out = fmemopen(buffer, EXTERN_CALL_BUFFER_SIZE, "w+");
```
Just noticed that, unfortunately, `fmemopen` doesn't work on Windows. Maybe passing around a FILE* isn't a good idea.

Maybe passing a sort of string builder, basically a `VEC` of `char` with specific functions for adding formatted strings. But I'm open to other ideas :^).

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -201,14 +213,38 @@ DO_CALLS_FUNC_SIGNATURE(do_calls_call) {
 
 DO_CALLS_FUNC_SIGNATURE(do_calls_text) {
+	// TODO better handle this unused
+	(void) ctx;
```
I don't think there is a better way to mark a variable unused, what you can do is create a macro:

```
#define UNUSED(var) (void)(var)
```

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -231,28 +267,77 @@ DO_CALLS_FUNC_SIGNATURE(do_calls) {
+	}
+	int i = 0;
+	while (loaded_libs[i].dl_handle != NULL){
+  	unload_lib(loaded_libs[i]);
```
wrong indentation

Grillo-0 commented on subprojects/vinumc/vinumc.c:
```diff
@@ -49,6 +50,8 @@ struct flag {
 		bool *boolean;
 		char **str;
+		char ***str_array;
+		size_t *count;
```
You need to put these fields into a struct, if you don't do that, both fields will be at the same address

Grillo-0 commented on subprojects/vinumc/vinumc.c:
```diff
@@ -107,6 +111,19 @@ static void parse_cmdline(const int argc, char **argv, struct ctx *ctx, struct f
+					    *f->ref_as.str_array = calloc(1, sizeof(char *));
+					} else {
+					    *f->ref_as.str_array = realloc(*f->ref_as.str_array, 
+					        (*f->ref_as.count + 1) * sizeof(char *));
```
wrong identation

Grillo-0 commented on subprojects/vinumc/testlib.c:
```diff
@@ -0,0 +1,33 @@
+#include <stdio.h>
```
move this file into the `tests` dir

Grillo-0 commented on subprojects/vinumc/testlib.c:
```diff
@@ -0,0 +1,33 @@
+#include <stdio.h>
+#include "extern_library.h"
+
+char* day_of_week(char* arg) {
```
as this lib is just for tests we don't need a complex function with potential bugs, use simpler function

Grillo-0 commented on docs/library_development.md:
```diff
@@ -0,0 +1,25 @@
+    {"function2_name", function2_pointer}, 
+    {}};     
+  return lib;                                                  
+}
```
some spaces at the end of line, remove them

Grillo-0 replied:
The spaces are here still 
![image](https://github.com/user-attachments/assets/61483021-8463-4d56-9b90-625327434bc2)

Grillo-0 commented on subprojects/vinumc/extern_library.c:
```diff
@@ -0,0 +1,21 @@
+	void *handle = dlopen(name, RTLD_LAZY);
+	if (!handle) {
+		// TODO: deal with the error of not being able to open the library
+	}
```
keep the TODO but change the if to an `assert(handle)`

artP2 committed:
vinumc/str.h: create a string builder
This adds a str definition and a function to append a str.

Grillo-0 commented on subprojects/vinumc/vec.h:
```diff
@@ -33,4 +33,17 @@
+		strcat((arr)->base, str);                                                          \
+		(arr)->len += _len;                                                                \
+	} while (0)
+
```
This only makes sense for a vec of chars. It would be better if we create a string builder in a new separate file and add this function to it, the string builder would just be a vec of chars.

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -183,14 +198,15 @@ RESOLVE_FUNC_SIGNATURE(resolve_calls) {
 }
 
 #define DO_CALLS_FUNC_SIGNATURE(func_name)                                                         \
-	static void func_name(const struct ast *ast, FILE *out, size_t ast_node_id)
+	static void func_name(struct eval_ctx *ctx, const struct ast *ast, struct str *out,        \
+			      size_t ast_node_id)
 
```
The change from a `FILE *` to a `struct str*` is a change that deserves a separate commit.

Grillo-0 replied:
Also, adding the `eval_ctx` could be a separate commit

Grillo-0 commented on subprojects/vinumc/eval.h:
```diff
@@ -4,13 +4,25 @@
+#include "extern_library.h"
 #include "vec.h"
 
+#define UNUSED(var) (void)(var)
```
I don't think that this `#define` makes sense been here, place it on `eval.c` or create an `utils.h` file.

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -201,14 +217,33 @@ DO_CALLS_FUNC_SIGNATURE(do_calls_call) {
+			find_symbol_on_scopes(&ctx->scopes, curr_scope, symbol_node->text);
+
+		// put the extern function call return on the out str
+		VEC_PUT_STR(out, symbol_info->as.func(tmp_out.base));
```
free the tmp_out

artP2 committed:
vinumc/utils.h: add the unused macro

artP2 committed:
vinumc/eval.c: add eval_ctx to do_call functions

artP2 committed:
vinumc/eval.c: change from FILE to str on do_call functions

artP2 committed:
vinumc/vec.h: add VEC_FREE and VEC_PUT_MANY macro

JeanJPNM commented on subprojects/vinumc/eval.c:
```diff
@@ -201,14 +217,34 @@ DO_CALLS_FUNC_SIGNATURE(do_calls_call) {
+
+		// put the extern function call return on the out str
+		PUT_STR(out, symbol_info->as.func(tmp_out.base));
+		VEC_FREE(&tmp_out);
```
This can potentially be a source of memory leaks if the function allocates its own string and returns it, making it the compiler's job to free it.

monacofj commented:
@Grillo-0 and I were discussing the fix to the free-non-allocated-memory problem at `vinumc/eval.c:239`

```
               if (call_return != tmp_out.base) {
			free(call_return);
		}
```

As far as I understood, the issue is whether to free or not the return value of a function, without knowing whether this value is pointer to either a dynamically or a statically allocated memory block.

My suggestion for this is to add one more piece of information in `struct extern_function` at `extern_library.h`

```
enum mem_type_t {mem_stat, mem_dyn};
struct extern_function {
        char *name;
        extern_function_pointer fp;
        mem_type_t memory;
};
```
that the function implementer should provide in `expose_library`, depending on whether they return a pointer to either a static or a dynamically allocated buffer. 

@Grillo-0 is more towards an alternativa approach: to change the functions API such that, instead of returning a `char *`, all functions would return a struct containing both the `char *` and a second member such as `mem_type_t`.

I myself prefer the first option: keep functions returning `char *` and having the user to specify whether the memory is static or dynamic when registering the function with `expose_library`.

An attempt in #52

JeanJPNM commented:
The current implementation does not provide a way to access symbols from the call scope (e.g. `[my_impl [name: John] [status: Garbage Disposal]]`). Will this be addressed separatedly in another pr?

Grillo-0 commented:
> The current implementation does not provide a way to access symbols from the call scope (e.g. `[my_impl [name: John] [status: Garbage Disposal]]`). Will this be addressed separatedly in another pr?

Yeah, let's leave this for another PR.

This one is alive for long enough XD.

Created #57 for that

Grillo-0 reviewed (approved):
Finally, it's done. I will just add @JeanJPNM credit on the commit message.

Tomorrow, it will be merged
